ceph.spec.in

```diff
@@ -1289,6 +1289,7 @@ install -m 0644 -D src/logrotate.conf %{buildroot}%{_sysconfdir}/logrotate.d/cep
 chmod 0644 %{buildroot}%{_docdir}/ceph/sample.ceph.conf
 install -m 0644 -D COPYING %{buildroot}%{_docdir}/ceph/COPYING
 install -m 0644 -D etc/sysctl/90-ceph-osd.conf %{buildroot}%{_sysctldir}/90-ceph-osd.conf
+install -m 0644 -D src/tools/rbd_nbd/rbd-nbd_quiesce %{buildroot}%{_sysconfdir}/ceph/rbd-nbd_quiesce
```

Comment:
Wrong permissions for a script. Also probably should live in "/usr/bin" like other helper scripts (i.e. ceph-rbdnamer). I guess this was probably so the user could tweak it, but since `rbd-nbd` now supports the `--queue-hook` optional, the end-user can create their own script anywhere.

Reply:
I suppose libexec directory ("/usr/libexec/rbd-nbd") would fit better for this type of scripts than "/usr/bin". Still I am not entirely sure it is better than "/etc/ceph". Yes, the reason is that the user (or rather administrator) could tweak it. Although `--quiesce-hook` is useful, the problem with this that the administrator can not enforce users to use it.

Consider a case, the administrator wants by some reason to disable quiesce (e.g. fsfreeze gets stuck frequently). With the current setup, she would just add `exit 0` as the first line to the script, so if a user run `rbd-nbd map --quiesce $image` the hook would be nop. If a user explicitly specifies `--quiesce-hook` than she is sure she wants some specific, non-default for this site, behavior. I think I would have liked to have the script in the libexec dir, instead of the  sysconfdir, if the rbd-nbd had had a config, where the administrator could specify the hook location. But we don't have such. In other words I would like the administrator has a possibility to configure the site default behavior for "quiesce", and it seems the only way to do this is to have the default `--quiesce-hook` location to be in the sysconfig ("/etc") dir.

It think having an executable script in '/etc' is rather common practice. Usually you can find many scripts there (init, network config, ...), and we already have rbdmap, which I actually used as an example for `rbd-nbd_quiesce`.

I think I would be fine to have rbd-nbd_quiesce in  the libexec dir and have "/etc/ceph/rbd-nbd_quiesce" as a symlink to it. Not sure though we really want this complication.

Reply:
All the NBD commands require `CAP_SYS_ADMIN` if I remember correctly, so it would not be a random user running `rbd-nbd`. You could add a config option for a default `quiesce-hook` location so that it can be overwritten globally, per-host, etc. 

Looks like we install `rbdmap` in `/usr/bin`. The new systemd scripts point to `/usr/bin`. The `rbdmap` that goes in `/etc/ceph` is the config file.

Reply:
A "user" could be just a third party software used to map rbd-nbd devices, which calls (may call) rbd-nbd map with `--quiesce` option but does not provide a possibility to pass `--quiesce-hook` option. 

> Looks like we install rbdmap in /usr/bin. The new systemd scripts point to /usr/bin. The rbdmap that goes in /etc/ceph is the config file.

Indeed. I just had a wrong impression about it then.

So you still prefer having it somewhere in bin or libexec? I am not going to insist. It is a little sad to me that it would not be possible to configure a system in such a way that a user could just call `rbd-nbd map --quiesce` and should always add `--quiesce-hook` if they want to use their own hook. But I can live with this -- anyway most likely user will not call the command directly but use some third party tool. (And having the hook in /etc has its own drawbacks: e.g. to make sure it is not overwritten on the upgrade).

Reply:
> So you still prefer having it somewhere in bin or libexec?

Yes, that is what I would lean towards. If you then add a config option for the default hook location, it seems like you could then configure a node to always use a user-provided hook just as easily as if the user tweaked the package-provided hook.

Reply:
Ok. I will move the hook to libexec. Not sure about the config option though (I suppose you mean ceph.conf, not a separate nbd config), because before this we have not have any config options for nbd and I am not sure we really want (mostly the same arguments as for using internal librbd functions).

Reply:
Yeah, I was talking about a `common/config.cc` addition to support MGR-based global config overrides + `ceph.conf` + CLI processing. I am not really pushing for it, but it offers a way to have a default hook override location and natively handles a hook override w/o the need to tweak `rbd device map`.

Reply:
I have changed the default hook location to be in  the rbd-nbd libexec directory. 

I have not added the config option though. It is not that I am against it, I am just reluctant to open this can of worms right now: if we add an rbd-nbd option for this, what about other rbd-nbd options? It would be logical to add them too. And if we add nbd options, what about kernel options, and ggate options? It would be logical to add them too...

So it can be done later together with other options or may be we think out some other solution.

Reply:
:+1: